pennylane/tape/tapes/jacobian_tape.py

```diff
@@ -239,22 +239,69 @@ def numeric_pd(self, idx, device, params=None, **options):
         shift[idx] = h
 
         if order == 1:
-            # Forward finite-difference.
-            # Check if the device has already be pre-computed with
-            # unshifted parameter values, to avoid redundant evaluations.
+            # forward finite-difference.
+
+            tapes = []
+
+            # get the stored result of the original circuit
             y0 = options.get("y0", None)
 
+            shifted = self.copy(copy_operations=True, tape_cls=QuantumTape)
+            shifted.set_parameters(params + shift)
+
+            tapes.append(shifted)
+
             if y0 is None:
-                y0 = np.asarray(self.execute_device(params, device))
+                tapes.append(self)
+
+            def processing_fn(results):
+                """Computes the gradient of the parameter at index idx via first-order
+                forward finite differences.
 
-            y = np.asarray(self.execute_device(params + shift, device))
-            return (y - y0) / h
+                Args:
+                    results (list[real]): evaluated quantum tapes
+
+                Returns:
+                    array[float]: 1-dimensional array of length determined by the tape output
+                    measurement statistics
+                """
+                shifted = np.array(results[0])
+                unshifted = y0
+
+                if unshifted is None:
+                    unshifted = np.array(results[1])
+
+                return (shifted - unshifted) / h
+
+            return tapes, processing_fn
 
         if order == 2:
             # central finite difference
-            shift_forward = np.array(self.execute_device(params + shift / 2, device))
-            shift_backward = np.array(self.execute_device(params - shift / 2, device))
-            return (shift_forward - shift_backward) / h
+
+            shifted_forward = self.copy(copy_operations=True, tape_cls=QuantumTape)
+            shifted_forward.set_parameters(params + shift / 2)
+
+            shifted_backward = self.copy(copy_operations=True, tape_cls=QuantumTape)
+            shifted_backward.set_parameters(params - shift / 2)
+
+            tapes = [shifted_forward, shifted_backward]
+
+            def second_order_processing_fn(results):
+                """Computes the gradient of the parameter at index idx via second-order
+                centered finite differences.
+
+                Args:
+                    results (list[real]): evaluated quantum tapes
+
+                Returns:
+                    array[float]: 1-dimensional array of length determined by the tape output
+                    measurement statistics
+                """
+                res0 = np.array(results[0])
+                res1 = np.array(results[1])
+                return (res0 - res1) / h
```

Comment:
No, should be h here, because the shift is +- h/2 😆

I have to admit, I did a double take, since I'm so used to seeing 2h in the denominator of centered finite differences.